File: curved_scales.py
import math
import pygame

WHITE = (255, 255, 255)


def _point_on_arc(cx: int, cy: int, radius: int, angle_deg: float) -> tuple[int, int]:
    # 0 deg = straight up, negative = left, positive = right
    a = math.radians(angle_deg)
    x = cx + radius * math.sin(a)
    y = cy - radius * math.cos(a)
    return int(x), int(y)


def draw_roll_marks(
    surf: pygame.Surface,
    horizon_left: int,
    horizon_top: int,
    horizon_width: int,
    horizon_height: int,
    colour: tuple[int, int, int] = WHITE,
) -> None:
    """
    Top roll scale:
    - apex touches the top of the blue horizon
    - arc spans full horizon width
    """
    cx = horizon_left + horizon_width / 2
    apex_y = horizon_top

    # How far lower the arc is at left/right edges than at center.
    # Increase for more curve, decrease for flatter.
    edge_drop = 18.0

    half_span = horizon_width / 2.0

    # Circle from chord/segment geometry
    radius = (half_span * half_span + edge_drop * edge_drop) / (2.0 * edge_drop)
    cy = apex_y + radius + 10

    # Angle where arc hits left/right horizon edges
    end_deg = math.degrees(math.asin(half_span / radius))

    start_i = int(-end_deg)
    end_i = int(end_deg)

    arc_points = [_point_on_arc(int(cx), int(cy), int(radius), d) for d in range(start_i, end_i + 1)]
    #pygame.draw.lines(surf, colour, False, arc_points, 2)

    # Static roll tick marks across the arc
    tick_fracs = [-1.0, -0.75, -0.5, -0.33, -0.16, 0.0, 0.16, 0.33, 0.5, 0.75, 1.0]
    for f in tick_fracs:
        ang = f * end_deg
        tick_len = 14 if abs(f) in (1.0, 0.5, 0.0) else 9
        p_outer = _point_on_arc(int(cx), int(cy), int(radius), ang)
        p_inner = _point_on_arc(int(cx), int(cy), int(radius - tick_len), ang)
        pygame.draw.line(surf, colour, p_outer, p_inner, 2)


def draw_heading_marks(
    surf: pygame.Surface,
    circle_cx: int,
    circle_cy: int,
    circle_radius: int,
    colour: tuple[int, int, int] = WHITE,
) -> None:
    # Bottom curved heading marks on the grey arch
    arc_points = [_point_on_arc(circle_cx, circle_cy, circle_radius, d) for d in range(-14, 15, 1)]
    pygame.draw.lines(surf, colour, False, arc_points, 2)

    for ang in range(-12, 13, 3):
        tick_len = 12 if ang % 6 == 0 else 7
        p_outer = _point_on_arc(circle_cx, circle_cy, circle_radius, ang)
        p_inner = _point_on_arc(circle_cx, circle_cy, circle_radius - tick_len, ang)
        pygame.draw.line(surf, colour, p_outer, p_inner, 2)

File: esp32_version/car_pfd_Banding/curved_scales.cpp
#include "curved_scales.h"
#include <math.h>

static void pointOnArc(int cx, int cy, int radius, float angleDeg, int &outX, int &outY) {
    float a = angleDeg * PI / 180.0;
    outX = cx + (int)(radius * sin(a));
    outY = cy - (int)(radius * cos(a));
}

void drawRollMarks(TFT_eSPI &tft, int horizonLeft, int horizonTop, int horizonWidth, int horizonHeight, uint16_t colour) {
    float cx = horizonLeft + horizonWidth / 2.0;
    float apexY = horizonTop;
    float edgeDrop = 18.0;
    float halfSpan = horizonWidth / 2.0;
    float radius = (halfSpan * halfSpan + edgeDrop * edgeDrop) / (2.0 * edgeDrop);
    float cy = apexY + radius + 10;
    float endDeg = asin(halfSpan / radius) * 180.0 / PI;

    float tickFracs[] = {-1.0, -0.75, -0.5, -0.33, -0.16, 0.0, 0.16, 0.33, 0.5, 0.75, 1.0};
    for (int i = 0; i < 11; i++) {
        float f = tickFracs[i];
        float ang = f * endDeg;
        int tickLen = (fabs(f) == 1.0 || fabs(f) == 0.5 || f == 0.0) ? 14 : 9;
        int outerX, outerY, innerX, innerY;
        pointOnArc((int)cx, (int)cy, (int)radius, ang, outerX, outerY);
        pointOnArc((int)cx, (int)cy, (int)(radius - tickLen), ang, innerX, innerY);
        tft.drawLine(outerX, outerY, innerX, innerY, colour);
    }
}

void drawHeadingMarks(TFT_eSPI &tft, int circleCx, int circleCy, int circleRadius, uint16_t colour) {
    int prevX, prevY;
    pointOnArc(circleCx, circleCy, circleRadius, -14, prevX, prevY);
    for (int d = -13; d <= 14; d++) {
        int x, y;
        pointOnArc(circleCx, circleCy, circleRadius, d, x, y);
        tft.drawLine(prevX, prevY, x, y, colour);
        prevX = x;
        prevY = y;
    }
    for (int ang = -12; ang <= 12; ang += 3) {
        int tickLen = (ang % 6 == 0) ? 12 : 7;
        int outerX, outerY, innerX, innerY;
        pointOnArc(circleCx, circleCy, circleRadius, ang, outerX, outerY);
        pointOnArc(circleCx, circleCy, circleRadius - tickLen, ang, innerX, innerY);
        tft.drawLine(outerX, outerY, innerX, innerY, colour);
    }
}